---
title: "RNASeq - Practical Session 2"
date: "7/24/2017"
output: 
  html_document:
    toc: true
    theme: spacelab
---

# Introduction
In contrast to differential expression analysis where differences in gene expression between experimental groups are compared, time series differential expression analysis also compares significant differences between experimental groups but also takes into account the effect different timepoints might have. In this way you can for example, identify genes that not only have an effect on a disease but also on the course of the disease. 
http://bioconductor.org/packages/release/bioc/html/maSigPro.html

## Install and load the library and data
```{r, message=FALSE}
# Install the library
source("https://bioconductor.org/biocLite.R")
biocLite("maSigPro")
#install.packages("pheatmap")
#install.packages("MASS")
library("pheatmap")
library("MASS")
library("maSigPro")

# Load the accompanying data
data(data.abiotic)
data(edesign.abiotic)
```

For this section we will use a public data set from a plant abiotic stress study performed at the TIGR Institute by (Rensink et al., 2005). In this study, potato plants were subjected to three different types of abiotic stresses (cold, heat and salt) and gene expression was monitored at three time points after the start of the treatments. RNA was also collected from non-stressed plants at the same time points and all samples where hybridased against a common control on a 11K cDNA potato chip. There are three biological replicates for each experimental condition. The matrix 'edesign.abiotic' contains the description of the experimental design. The dataframe 'data.abiotic' contains normalized expression data for a random set of 1000 genes of this study. 

# Building the regression model

## Defining the models
```{r}
design <- make.design.matrix(edesign.abiotic, degree = 2)
```

The first task is definining which models need to be evaluated. This is done using the function 'make.design.matrix' and the experimental desing matrix 'edesign.abiotic'. The degree represents the degrees of the polynome, i.e. which type of model to fit. Since the data consists of 3 time points a quadratic model can be used. This is represented by a 2. 

Q1: Try also building the design matrix with a degree of 1 and see the difference

## Finding significant genes
```{r}
fit <- p.vector(data.abiotic, design, Q = 0.05, MT.adjust = "BH", min.obs = 20)
```

The function 'p.vector' takes the previously constructed design object and the gene expression data and builds the regression models and calculates their significance. The parameter 'Q' represents the p-value threshold at which a model is considered to be significant. 'MT.adjust' specifies the method of accounting for multiple testing error (type 1 error rate), in this case the Benjamini-Hochberg method. Min.obs specifies how many samples must be present for the model to be fitted. If not enough samples are present for a gene a reliable model can not be built, so genes with many missing samples must be omitted.

Q2: How many genes show significance for the specified theshold Q = 0.05?

Q3: And for Q = 0.005?

Q4: What is the effect on the overall amount of signifcant genes when no multiple testing correction is applied?

Stuck? Try reading the help using ?p.vector

## Finding differences between experimental conditions

```{r}
tstep <- T.fit(fit, step.method = "backward", alfa = 0.05)
```

Next selection of which variables are significant is perfomred. This is done using a process called stepwise regression. During this process variables are added to a null model (forward) or removed from the full model (backward) one by one and at each step the model quality is assesed. In this case the backward selection method is used so it starts of with a model with all variables and they are subsequently removed one by one. The parameter alfa represents the p-value threshold variables must have to be included in the model.

Q5: What would be the effect of increasing alfa?

## Getting the list of significant genes

```{r}
sigs <- get.siggenes(tstep, rsq = 0.6, vars = "groups")
```

This function selects all genes/models above a certain r-squared threshold (the amount of variance the model for each gene could explain) for all experimental groups (as defined by the vars argument). 

# Visualizing the results
```{r}
suma2Venn(sigs$summary[, c(2:4)])
```

This venn diagram shows the overlap in genes between the experimental conditions

Q6: How many genes show significance between all of the timepoints but also the experimental groups?

Q7: What percentage of the genes is unique to a single experimental condition?

```{r, message=FALSE, results='hide'}
pdf(width=20, height=10, file="SeeGenesOutput.pdf")
see.genes(sigs$sig.genes$ColdvsControl,
          show.fit = T,
          dis = design$dis,
          cluster.method = "hclust",
          cluster.data = 1,
          newX11 = F,
          k = 9,
          cexlab = 2)
dev.off()
```

This function plots 2 plots. Firstly an overview of the gene clusters. Secondly it plots the overall expression of a gene cluster for each experimental condition over time. To view the plots open the pdf 'SeeGenesOutput.pdf' in your working directory.

```{r}
pheatmap(as.matrix(na.omit(data.abiotic[as.character(sigs$summary$ColdvsControl),])))
```

We can also plot a heatmap to visualize the differences between the experimental groups and replicates. In this case we are looking at the significantly regulated genes for cold over time.

Q8: Do the observed clusters in the heatmap match your expectations?

Q9: Do the results shown in the heatmap match with the previously created venn diagram?

Q10: Do the same for some of the other experimental groups such as 'SaltvsControl'

# Using count data
The above example was done using microarray data but the same analysis can also be performed using RNAseq. Using the following data and the above examples implement the analysis yourself with the RNAseq data. One thing to look out for is to specify in 'p.vector' that count data is being used with the option 'counts=TRUE'. The recomended cluster size is k=4 for this dataset, but feel free to vary this to see the changes. If you have time try to edit the tresholds mentioned above and observe the changes

```{r}
data(NBdata)
data(NBdesign)
```

## Put your code here
```{r}

```